src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.cs

```diff
@@ -0,0 +1,170 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json.Serialization
+{
+    public static partial class JsonConverter
+    {
+        private static readonly JsonConverterSettings s_default_options = new JsonConverterSettings();
+
+        private static bool ReadData(
+            ref Utf8JsonReader reader,
+            JsonConverterSettings options,
+            Type returnType,
+            ref JsonObjectState current,
+            ref List<JsonObjectState> previous,
+            ref int arrayIndex)
+        {
+            while (reader.Read())
+            {
+                JsonTokenType tokenType = reader.TokenType;
+                if (tokenType == JsonTokenType.String)
+                {
+                    CheckForNonObjectReturn(ref current, options, returnType);
+                    if (HandleString(ref reader, options, returnType, ref current))
+                    {
+                        return true;
+                    }
+                }
+                else if (tokenType == JsonTokenType.Number)
+                {
+                    CheckForNonObjectReturn(ref current, options, returnType);
+                    if (HandleNumber(ref reader, options, returnType, ref current))
+                    {
+                        return true;
+                    }
+                }
+                else if (tokenType == JsonTokenType.PropertyName)
+                {
+                    Debug.Assert(current.ReturnValue != default);
+                    Debug.Assert(current.ClassInfo != default);
+
+                    ReadOnlySpan<byte> propertyName = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan;
```

Comment:
If your `Utf8JsonReader` was known to be constructed from a byte[] or span (rather than `ReadOnlySequence`), we can get rid of the `reader.HasValueSequence` check.

Reply:
Hmm not sure if passing that info down is any better\quicker

Reply:
What info would you have to pass down? If this is a general purpose work-horse which needs to work for any instance of `Utf8JsonReader`, **then keep the code as is**. What you have now is a more robust implementation, so I am fine with keeping it.

If on the other hand, we only every construct it from spans (or single-segment ReadOnlySequence), the value will always be contained in `ValueSpan` so checking and calling ToArray on `ValueSequence` is pointless. I was just pointing out that we could leverage the info about how `Utf8JsonReader` was constructed since we control it.